src/App/Utilities/MapToolExporter.ts

```diff
@@ -0,0 +1,293 @@
+import { OrderedMap } from "../../Data/OrderedMap"
+import { ActivatableDependent } from "../Models/ActiveEntries/ActivatableDependent"
+import { HeroModel } from "../Models/Hero/HeroModel"
+import { Record } from "../../Data/Record"
+
+export function getPropertiesXML (): string {
+  return "<?xml version=\"1.0\"?><map><entry><string>version</string><string>1.7.0</string></entry>"
+  + "<entry><string>herolab</string><boolean>false</boolean></entry></map>"
+}
+
+function separateValuesByComma (map: OrderedMap<string, Record<ActivatableDependent>>): string {
+  let text = ""
+  map.value.forEach ((value, key, map) => {
+    text += ` ${value.name},`
+  })
```

Comment:
Es so zu machen ist meinetwegen ok, allerdings kann es natürlich sein, dass sich interne Darstellungen ändern, dann hättest du hier ein Problem. Es wäre besser, wenn du die dafür vorgesehenen Funktionen aus dem Modul nimmst. Du könntest sowas machen wie:

```suggestion
  pipe_ (
    map,
    OrderedMap.elems,
    Maybe.mapMaybe(pipe (ActivatableDependent.A.id, OrderedMap.lookupF (staticMap), fmap(getName)))),
    List.intercalate(", ")
  )
```

Wobei du hier `staticMap` und `getName` noch übergeben müsstest, je nachdem, welchen Typ Entität du haben möchest. Die `ActivatableDependent`-Records halten nur die charakterspezfisichen, "dynamischen" Werte, die statischen Werte aus der Datenbank sind in separaten Records.

Reply:
Und wo bekomme ich diese staticMap überhaupt her?

Reply:
Daran kommst du über `getWiki` aus `stateSelectors.ts`, der du den AppState übergeben musst.

Reply:
Ich habe das ganze etwas umstrukturiert. Sieht jetzt so aus:

`  xml += entry ({
    key: "Vorteile",
    value: pipe_(
      hero.values.advantages,
      OrderedMap.elems,
      Maybe.mapMaybe (pipe_ (ActivatableDependent.A.id.toString(),
                             OrderedMap.lookupF (getWiki()),
                             fmap((adv)=>{adv}))), //hier zeigt er mir keine Properties
      List.intercalate(", "))
    })`

Irgendwo ist noch ein Typfehler drin. Ich blick da aber auch nicht wirklich durch. Keine Ahnung was hier passiert.

Reply:
- Also `pipe_` muss `pipe` sein, weil du einen Parameter aus `mapMaybe` bekommst.
- Bei `Activatable.A.id` darfst du kein `toString()` machen, weil die Funktion ja die ID aus dem Parameter lesen muss.
- `getWiki()` müsste `getWikiAdvantages (state)` sein. 
- `(adv) => {adv}` gibt `undefined` zurück, wenn du ein Objekt mit der Eigenschaft `adv` zurückgeben willst, musst du `(adv) => ({adv})` schreiben. 
- `List.intercalate` ist quasi `Array#join`. Das geht hier nicht, weil du ein Objekt zurückgibst, das einen Record enthält.

Was das ganze machen *soll*: 
Du möchtest ja die Namen für alle Vorteile bekommen. Bei einem Charakter werden aber nur die aktuellen Zustände der Vorteile gespeichert, nicht die Vorteile selbst. Daher musst du dir über die ID erst einmal den dazugehörigen Vorteil holen. Wenn es den gibt, kannst du dir z. B. per `Advantage.A.name` den Namen extrahieren – also in das `fmap` einsetzen. Damit hättest du eine Liste von Strings, die du dann mit `List.intercalate` joinen kannst.

Das generiert dir aber nur die Namen. Wenn du die Auswahloptionen auch mit ausgeben möchtest, kommst du damit nicht weit. In dem Fall würde ich dir empfehlen, mit `getAdvantagesForSheet` aus `activatableSelectors.ts` zu arbeiten. Das Ergebnis kannst mit `compressList (staticData) (list)`, wobei `list` das Ergebnis aus dem zurückgegebenen `Maybe` ist, als einzelnen String formatieren und dann speichern.

Reply:
Also ich hab jetzt mal lange genug rumgespielt um das System der Funktionen zu verstehen. Ich hab das ganze jetzt mal aufgeteilt um die Typen ein bischen genauer analysiert die man hier braucht. Aktuell will ich mal nur den Namen herausfinden und stoße auf ein kleines Problem:

```
  const getAdvantageName = pipe (
    ActivatableDependent.A.id,
    OrderedMap.lookupF (getWikiAdvantages(state)),
    fmap ((adv: Record<Advantage>)=>{
      adv.values.name
    }))

  xml += entry ({
    key: "Vorteile",
    value: pipe_(
      hero.values.advantages,
      OrderedMap.elems, //gibt mir List<Record<ActivatableDependent>>
      fmap(getAdvantageName), //will Functor<Record<ActivatableDependent>>
      List.intercalate(", "))
    })
```

Reply:
Der untere Teil müsste so sein:

```ts
  xml += entry ({
    key: "Vorteile",
    value: pipe_(
      hero.values.advantages,
      OrderedMap.elems, 
      Maybe.mapMaybe(getAdvantageName), // mapMaybe statt fmap
      List.intercalate(", "))
    })
```

Dann sollte er eigentlich keine Fehler schmeißen …

Reply:
Das hat nichts geändert. Immer noch der gleiche Typfehler.
![grafik](https://user-images.githubusercontent.com/765587/101235156-41c84300-36c6-11eb-99a1-f538b10caf6c.png)

Reply:
So wie ich es verstanden habe müsste Maybe.mapMaybe ja das gleiche machen wie fmap nur dass die Maybe-Elemente, die nichts enthalten, quasi bei der Ausführung rausgefiltert werden. Und da ich von OrderedMap.elems keine Maybes bekomme dürfte das keinen Unterschied machen

Reply:
Aber `getAdvantageName` gibt ein `Maybe` zurück. Und das müsste ja herausgefiltert werden. Ich frage mich gerade nur, wieso er den Typ nicht richtig inferieren kann … 🤔

Reply:
Statt 

```
(adv: Record<Advantage>)=>{
      adv.values.name
    })
```

würde ich sonst mal 

```
Advantage.A.name
```

in `fmap` einsetzen. Ich habe eben gesehen, dass als Rückgabetyp von `getAdvantageName` offenbar `Maybe<void>` inferiert wird, was ja nun nicht wirklich stimmt.

Reply:
Die Kürzung hat funktioniert. getAdvantageName gibt durch die Änderung Maybe<string>zurück. Jetzt bekomme ich hier zumindest keinen Kompilierfehler mehr.